### PR TITLE
localisationTestForInvalidStrings

### PR DESCRIPTION
**Pull request by steph37tours**
…te for instance

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Screenshots added (for bigger UI changes)
- [ ] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)
- [ ] If you changed the localization: Did you run `gradle localizationUpdate`?

### CODE REVIEW

**Comment by lynyus** [comment]
Hey @steph37tours, this goes into the right direction. I think the test fails in the right places, but the message is quite big.
What would be better is something like this:
`Found an invalid character in the <language> localization of <JabRef | Menu>: The <key | value> <actual string> contains a space!`

This would require to iterate over all property strings, similar as in `LocalizationConsistencyTest.keyValueShouldBeEqualForEnglishPropertiesMenu()`.
Please give it another shot.

**Comment by steph37tours** [comment]
Thank you @Lynyus. I see what you mean. I'll correct this tomorrow if it's ok :-).

**Comment by koppor** [comment]
I am unsure whether JabRef really cannot handle underscores. I once looked at the code and I think, we could get rid of using the underscores and always use spaces. Could you check that? I think `_` is so 2005 😇

**Comment by lynyus** [comment]
Wow @koppor (;

**Comment by lynyus** [comment]
@koppor to make this crystal clear: Do we need to fix the underscores or not? 
In https://github.com/JabRef/jabref/blob/ebbeb1d249edf7134abf92d24c2031ae54d19881/src/main/java/org/jabref/logic/l10n/LocalizationKey.java#L14 there is a comment that fixes the spaces, but not the `#`, `!`, `=`, and `:`.
So is there need to pursue this PR, or not?

**Comment by Siedlerchr** [comment]
@koppor  @lynyus  That is specified in the java docs:
http://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#load(java.io.Reader)

> .A comment line has an ASCII '#' or '!' as its first non-white space character; comment lines are also ignored and do not encode key-element information
In addition to line terminators, this format considers the characters space (' ', '\u0020'), tab ('\t', '\u0009'), and form feed ('\f', '\u000C') to be white space. 

> The key contains all of the characters in the line starting with the first non-white space character and up to, but not including, the first unescaped '=', ':', or white space character other than a line terminator

**Comment by koppor** [comment]
Then, we can have the values without underscores? That would improve the language files a lot!

**Comment by lynyus** [comment]
The question is, of course, why they were introduced in the first place. Is there any pitfall if we remove them?!?

**Comment by Siedlerchr** [comment]
From what I understand the key can't contain whitespaces or it would be recognized as key:
`A key is all the characters up to the first whitespace or a key/value separator `
http://kajabity.com/kajabity-tools/java-properties-classes/the-java-properties-file-format/

**Comment by steph37tours** [comment]
I just push a change. I hope it is what you want. I've got difficult to test it.
And I'm not sure for the "throws IOException" Eclipse ask  me for remove it.
Thank you for help and permit to participate. It was very interesting for me.

File: src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java [hunk 1]
```diff
@@ -183,6 +183,28 @@ public void localizationParameterMustIncludeAString() throws IOException {
         }
     }
 
+    @Test
+    public void localisationTestForInvalidStrings() throws IOException {
```
Comment (lynyus) [inline src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: Indeed, you can remove the throws statement.

File: src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java [hunk 2]
```diff
@@ -183,6 +183,28 @@ public void localizationParameterMustIncludeAString() throws IOException {
...
+                String propertyFilePath = String.format("/l10n/%s_%s.properties", bundle, lang);
+
+                // read in
+                Properties textKeys = LocalizationParser
+                        .getProperties(propertyFilePath);
+                //parse object "textKeys" to find any spaces
+                for (Map.Entry<Object, Object> entry : textKeys.entrySet()) {
+                    String expectedKeyEqualsKey = String.format("%s=%s", entry.getKey(), entry.getKey());
+                    String actualKeyEqualsValue = String.format("%s=%s", entry.getKey(),
+                            entry.getValue().toString().replace(" ", ""));
```
Comment (lynyus) [inline src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: I would suggest that you check for spaces using String.contains(). Then there is no need to compare strings. later.

File: src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java [hunk 3]
```diff
@@ -183,6 +183,28 @@ public void localizationParameterMustIncludeAString() throws IOException {
...
+            for (String lang : Languages.LANGUAGES.values()) {
+                String propertyFilePath = String.format("/l10n/%s_%s.properties", bundle, lang);
+
+                // read in
+                Properties textKeys = LocalizationParser
+                        .getProperties(propertyFilePath);
+                //parse object "textKeys" to find any spaces
+                for (Map.Entry<Object, Object> entry : textKeys.entrySet()) {
+                    String expectedKeyEqualsKey = String.format("%s=%s", entry.getKey(), entry.getKey());
+                    String actualKeyEqualsValue = String.format("%s=%s", entry.getKey(),
```
Comment (lynyus) [inline src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: This is not correct. While in the english localization the key must equal the value, in all other languages the the key is different from the value.

File: src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java [hunk 4]
```diff
@@ -183,6 +183,28 @@ public void localizationParameterMustIncludeAString() throws IOException {
...
+                            entry.getValue().toString().replace(" ", ""));
+
+
+                    assertEquals("Found an invalid character in the "+ lang + "localization of" + bundle + ": The " + entry.getKey().toString() + ":" + expectedKeyEqualsKey + " contains a space!",expectedKeyEqualsKey, actualKeyEqualsValue);
```
Comment (lynyus) [inline src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: This looks fine, however make sure to include some spaces and quotes to make the sentence more readable
`Found an invalid character in the inlocalization ofJabRef: The pairs_processed:pairs_processed=pairs_processed contains a space! `
=>
`Found an invalid character in the 'in' localization of 'JabRef': The 'pairs_processed:pairs_processed=pairs_processed' contains a space! `

Also the error should be apparent from the message. The String above does not have a space.

Comment (steph37tours) [reply src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: I give you push access. I'm correcting my code.

Comment (steph37tours) [reply src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: I think it's ok now. But I didn't success to test ! It seems I don't know how to do.

File: src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java [hunk 5]
```diff
@@ -183,6 +183,23 @@ public void localizationParameterMustIncludeAString() throws IOException {
...
+        for (String bundle : Arrays.asList("JabRef", "Menu")) {
+            for (String lang : Languages.LANGUAGES.values()) {
+                String propertyFilePath = String.format("/l10n/%s_%s.properties", bundle, lang);
+
+                // read in
+                Properties textKeys = LocalizationParser
+                        .getProperties(propertyFilePath);
+                //parse object "textKeys" to find any spaces
+                for (Map.Entry<Object, Object> entry : textKeys.entrySet()) {
+                    assertTrue("Found an invalid character in the " + lang + " localization of " + bundle + " : The " + entry.getKey().toString() + " : " + entry.getValue().toString() + " contains a space!", entry.getValue().toString().contains(" "));
```
Comment (lynyus) [inline src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: The statement must be inverted. We want the test to fail if there is a '` `' in the String.
Also, you only verify the values, but not the keys of the entry.
Just add another line where you assert that there are also no spaces in the key.

File: src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java [hunk 6]
```diff
@@ -183,6 +183,23 @@ public void localizationParameterMustIncludeAString() throws IOException {
         }
     }
 
+    @Test
+    public void localisationTestForInvalidStrings() {
```
Comment (lynyus) [inline src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: I would suggest to make this name a bit more precise: maybe locali**z**ationTestForInvalidCharacters?

File: src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java (hunk 5 above)
Comment (steph37tours) [reply src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: @lynyus, I commit and push again. 
I correct the message with "At key", which seems to me more clear to find the line.
For the test : I use Junit. It find some failures

java.lang.AssertionError: Found an invalid character in the in localization of JabRef : proces_berpasangan At key : pairs_processed contains a space!

If I correct the P into Pairs_processed, it runs.

But I've got the same trouble in New_article , but I don't find why...

Comment (lynyus) [reply src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: Have you tried using assertFalse instead if assertTrue?

**Comment by lynyus** [comment]
So if you change the Assertions to assertFalse or negate the condition the tests should be operational and we can see if they pass the test on our build server
`assertFalse("Found an invalid character in the " + lang + " localization of " + bundle + " : " + entry.getValue().toString() + " At key : " + entry.getKey().toString() + " contains a space!", entry.getValue().toString().contains(" "));`

**Comment by steph37tours** [comment]
Yes, I change and commit. I found spaces in pt_BR localization :

java.lang.AssertionError: Found an invalid character in the pt_BR localization of JabRef : Normalizar para o formato_de_nome_do_BibTeX At key : Normalize_to_BibTeX_name_format contains a space!

**Comment by steph37tours** [comment]
So it is done now for me ?

**Comment by lynyus** [comment]
Can you push your commit please? (; You'll have to pull first, as I have already fixed the localization files on your branch.

File: src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java (hunk 5 above)
Comment (steph37tours) [reply src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: Yes. It runs !

**Comment by steph37tours** [comment]
sorry, it doesn't want to push ! It seems I have to pull your change before. I 'm looking at...

**Comment by steph37tours** [comment]
I'm sorry I can't push !
I've got the error that the branch countain work I doesn't have in local.
I try git pull
git pull --rebase without success for instance !

**Comment by lynyus** [comment]
on the current branch try the following:
`git fetch`
`git merge`
`git push`

**Comment by steph37tours** [comment]
git merge said not remote branch for the current

**Comment by steph37tours** [comment]
git merge Localisation_test_for_invalid_Strings 
said up-to-date

**Comment by lynyus** [comment]
Hm. Maybe try a GUI client like Github Desktop to figure everything out. Or clone the repository again and apply your changes again.

**Comment by steph37tours** [comment]
I'm sorry, I just clone and try to push again with the same error message. I continue to look for a solution.

**Comment by steph37tours** [comment]
I don't know what to do.
Into Eclipse, the menu team/pull give to me an error "Missing commit"

**Comment by steph37tours** [comment]
perhaps I have to do another fork ?

**Comment by steph37tours** [comment]
It seems it's ok now. I don't know what happen exactly.
Really thank you for this first experience, thank you for the time you share with me to help.

File: src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java [hunk 7]
```diff
@@ -183,6 +182,25 @@ public void localizationParameterMustIncludeAString() throws IOException {
...
+            for (String lang : Languages.LANGUAGES.values()) {
+                String propertyFilePath = String.format("/l10n/%s_%s.properties", bundle, lang);
+
+                // read in
+                Properties textKeys = LocalizationParser
+                        .getProperties(propertyFilePath);
+
+                //parse object "textKeys" to find any spaces
+                for (Map.Entry<Object, Object> entry : textKeys.entrySet()) {
+                    assertFalse("Found an invalid character in the " + lang + " localization of " + bundle + " : " + entry.getValue().toString() + " At key : " + entry.getKey().toString() + " contains a space!", entry.getValue().toString().contains(" "));
```
Comment (Siedlerchr) [inline src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: Just a minor thing: if you concatenate Strings and objects (no matter which type) with` "some text " + object    `
You don't need to do an explicit call to "toString()" The method is automatically invoked. But beware, in the case where you want to call a methond on a string like the contains, you have to use the toString

Comment (steph37tours) [reply src/test/java/org/jabref/logic/l10n/LocalizationConsistencyTest.java]: Thank you @Siedlerchr for this explanation.

**Comment by lynyus** [comment]
I'll merge this now, thanks for the efforts, @steph37tours! I hope you have learned something and are still eager to contribute. (;
How about this? https://github.com/JabRef/jabref/issues/2942 It's a small feature in the UI, and everybody would be quite happy if somebody finds time to implement it!

**Comment by steph37tours** [comment]
Thank you very much @lynyus !
I learn many things with this experience.
So I agree and wish to continue ! :-)

**Comment by koppor** [comment]
Good to see that you won the git war :). -- For learning git, I would recommend the tutorials listed at https://github.com/dictcp/awesome-git.

**Comment by steph37tours** [comment]
Thank you @koppor for the link ! I'll study it !

**Pull request merged**